# SUSA Kaggle Competition
## Part 1: Exploratory Data Analysis and Kaggle Introduction
### Hosted by and maintained by the [Statistics Undergraduate Students Association (SUSA)](https://susa.berkeley.edu). Originally authored by [Arun Ramamurthy](mailto:contact@arun.run), [Patrick Chao](mailto:prc@berkeley.edu), & [Noah Gundotra](mailto:noah.gundotra@berkeley.edu)


# Kaggle

## What is Kaggle?

## Machine Learning

## Accessing the `House Prices` Dataset

`crash-course/Kaggle/DATA/house-prices/train.csv`



# Exploratory Data Analysis

The first step is to load in the data! We will store this in a pandas dataframe.


In [69]:
import pandas as pd
train = pd.read_csv('DATA/house-prices/train.csv')
test = pd.read_csv('DATA/house-prices/test.csv')

#Let's see what the training dataframe looks like!
train.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities    ...     PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
5         Lvl    AllPub    ...            0    NaN  MnPrv        Shed     700   
6         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
7         Lvl    AllPub    ...            0    NaN    NaN        Shed     350   
8         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
9         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  
5     10   2009        WD         Normal     143000  
6      8   2007        WD         Normal     307000  
7     11   2009        WD         Normal     200000  
8      4   2008        WD        Abnorml     129900  
9      1   2008        WD         Normal     118000  

[10 rows x 81 columns]

## Questions for Understanding:
1. How many features do we have?
2. List 3 issues/questions that you see from the dataset
3. What are some important things we should do for data cleaning and exploration?

# The Data Science Workflow

In general, there are a few key steps to begin working with a dataset. First, we need to understand what the dataset actually is about, and what we are trying to do with it. Second it is very helpful to identify the size of the dataset, so we know how many samples we have. We need to determine a consistent method of dealing with missing values, such as setting them to a value, removing the feature entirely, interpolating values, etc. Other crucial steps are separating into training and validation, as well as creating elementary data plots. 

## What is our dataset? TODO
Our dataset is from blank, and we are trying to do blank 

## Data Cleaning
First, let's see how big our dataset looks like.

In [18]:
print("Training size",train.shape)
print("Test size",test.shape)

Training size (1460, 81)
Test size (1459, 80)


This tells us that we have $1460$ datapoints in the training set, and we would like to predict on $1459$ samples. There are $80$ total features, and one response variable we would like to predict. However, not all these 'features' are actually useful, such as `Id`. Thus we need to understand what the actual variables mean. A huge part in data science is actually understanding the variables. Of course it is possible to throw the data into some machine learning model and have it spit out predictions, but without actually understanding what data you are dealing with and how to feed the data into the model, your model is worthless.

Spend a good deal of time reading over the data dictionary. It is located in  
`crash-course/Kaggle/DATA/house-prices/data_description.txt`

## Questions for Understanding:
1. There are many categorical variables. What are some possibilities to deal with these variables? (Hint: [one hot encoding](https://en.wikipedia.org/wiki/one_hot_encoding))
2. Are there any categorical variables that we can convert to numerical/quantitative variables as well? How might we do that?
3. Are there any variables that are just irrelevant and we can ignore?

## Dealing with NA values
In almost all datasets, we will have NA values. These can be a pain to deal with, as there are many viable choices of what to do. First, it is good to see what columns have NA values.

In [20]:
#Sum the number of NA's in each column
train.isnull().sum()

Id                  0
MSSubClass          0
MSZoning            0
LotFrontage       259
LotArea             0
Street              0
Alley            1369
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType          8
MasVnrArea          8
ExterQual           0
ExterCond           0
Foundation          0
                 ... 
BedroomAbvGr        0
KitchenAbvGr        0
KitchenQual         0
TotRmsAbvGrd        0
Functional          0
Fireplaces          0
FireplaceQu       690
GarageType         81
GarageYrBlt        81
GarageFinish       81
GarageCars          0
GarageArea          0
GarageQual         81
GarageCond         81
PavedDrive

It seems that many of the features have a great deal of NAs. However, this is not necessarily the case. The astute reader will notice that some variables like `Alley`, `PoolQC`, and `Fence`, have NA has an actual value. For example for `Pool Quality`, we have these possibilities.

PoolQC: Pool quality
		
       Ex	Excellent
       Gd	Good
       TA	Average/Typical
       Fa	Fair
       NA	No Pool
So having NA values is not the usual not applicable (? forgot what NA stands for), it can actually be a legitimate value! We need to parse through the data dictionary to see when NA's are actually significant, and when they actually mean NA.

## Questions for Understanding:
1. Go through the data dictionary and find all features that have NA as legitimate values.
2. How can we refactor the variables in question 1 appropriately?


The features that we found with NA's are:
`Alley`, `BsmtQual`, `BsmtCond`, `BsmtExposure`, `BsmtFinType1`, `BsmtFinType2`, `FireplaceQu`, `GarageType`, `GarageFinish`, `GarageQual`, `GarageCond`, `PoolQC`, `Fence`, `Misc Feature`.

Alley: NA for no alley access  
`BsmtQual`, `BsmtCond`, `BsmtExposure`, `BsmtFinType1`, `BsmtFinType2`: NA for no basement  
`FireplaceQu`: NA for no fireplace  
`GarageType`, `GarageFinish`, `GarageQual`, `GarageCond`: NA for no garage  
`PoolQC`: NA for no pool  
`Fence`: NA for no fence  
`Misc Feature`: NA for no other miscellanous features (i.e. elevator, 2nd garage, shed, tennis Court, other)  

In [75]:
# Data Cleaning Functions


# This function takes in input as: 
# Data frame 'df'
# List of features to one hot encode 'features'
# Boolean for how to deal with NA's 'withNA'

# This function creates a set of output features from a categorical feature
# The output features are one hot encodings with names featureName_{value}
# Returns a new data frame (does not modify original dataframe) with appended features
# Usually the NA values are ignored and not considered one of the values
# If the boolean withNA is true, it will also add a column featureName_none one hot encoding of NA values

# Code from Numpy and Pandas SUSA guide
# crash-course/Python/Numpy and Pandas.ipynb
def oneHotFeature(df, features, withNA=False):
    #Copy over data
    newDf = df.copy()
    for feature in features:
        if withNA:
            newDf[feature] = newDf[feature].fillna('none')
        col_onehot = pd.get_dummies(newDf[feature], prefix=feature) 
        newDf.drop(feature, axis=1)
        newDf = newDf.join(col_onehot)
    return newDf

# This function takes in input as: 
# Data frame 'df'
# A single categorical feature to to be mapped 'feature'
# A mapping dictionary 'mapping'

# This function creates takes a dataframe and categorical feature, and maps the categorical values
# using the dictionary mapping
# Returns a new data frame (does not modify original dataframe) with modified values for the feature column
# For mappings with NA values, use 'NA' in the dictionary, this function properly deals with them
def categoricalToQuantitative(df,feature,mapping):
    newMapping = mapping.copy()
    newDf = df.copy()
    newDf[feature] = newDf[feature].fillna('none')
    if 'NA' in newMapping.keys():
        newMapping['none'] = newMapping['NA']
    newDf[feature] = newDf[feature].apply(lambda feat: newMapping[feat] if feat in newMapping.keys() else feat)
    return newDf



In [76]:
oneHotFeature(train,['Foundation']).head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities       ...        YrSold SaleType SaleCondition  \
0         Lvl    AllPub       ...          2008       WD        Normal   
1         Lvl    AllPub       ...          2007       WD        Normal   
2         Lvl    AllPub       ...          2008       WD        Normal   
3         Lvl    AllPub       ...          2006       WD       Abnorml   
4         Lvl    AllPub       ...          2008       WD        Normal   
5         Lvl    AllPub       ...          2009       WD        Normal   
6         Lvl    AllPub       ...          2007       WD        Normal   
7         Lvl    AllPub       ...          2009       WD        Normal   
8         Lvl    AllPub       ...          2008       WD       Abnorml   
9         Lvl    AllPub       ...          2008       WD        Normal   

  SalePrice Foundation_BrkTil Foundation_CBlock Foundation_PConc  \
0    208500                 0                 0                1   
1    181500                 0                 1                0   
2    223500                 0                 0                1   
3    140000                 1                 0                0   
4    250000                 0                 0                1   
5    143000                 0                 0                0   
6    307000                 0                 0                1   
7    200000                 0                 1                0   
8    129900                 1                 0                0   
9    118000                 1                 0                0   

   Foundation_Slab  Foundation_Stone  Foundation_Wood  
0                0                 0                0  
1                0                 0                0  
2                0                 0                0  
3                0                 0                0  
4                0                 0                0  
5                0                 0                1  
6                0                 0                0  
7                0                 0                0  
8                0                 0                0  
9                0                 0                0  

[10 rows x 87 columns]

In [82]:
mapping = {}
mapping['Ex'] = 5
mapping['Gd'] = 4
mapping['TA'] = 3
mapping['Fa'] = 2
mapping['Po'] = 1
mapping['NA'] = 0
print(train['FireplaceQu'].head(10))
print(categoricalToQuantitative(train,'FireplaceQu',mapping)['FireplaceQu'].head(10))

0    NaN
1     TA
2     TA
3     Gd
4     TA
5    NaN
6     Gd
7     TA
8     TA
9     TA
Name: FireplaceQu, dtype: object
0    0
1    3
2    3
3    4
4    3
5    0
6    4
7    3
8    3
9    3
Name: FireplaceQu, dtype: int64


In [ ]:
## EDA

```{r}
training_set <- read_csv("DATA/house-prices/train.csv") 
```


```{r}
training_set %>% summarise_all(funs(sum(is.na(.))/nrow(training_set))) %>% gather(Column, Count) %>% arrange(desc(Count)) %>% head(10)
eda_set <- training_set %>% drop_na(-c(PoolQC, MiscFeature, Alley, Fence, FireplaceQu, LotFrontage))
nrow(eda_set)/nrow(training_set)
```

```{r}
pairs_plot <- function(df){
  df_x <- rename_all(df, function(str){paste(str, "x", sep = ".")})
  df_y <- rename_all(df, function(str){paste(str, "y", sep = ".")})
  cbind(df_x,df_y) %>%
    gather(xVar, xVal, ends_with("x")) %>% gather(yVar, yVal, ends_with("y")) %>% 
    ggplot(aes(xVal, yVal, col = xVar)) + geom_point() + facet_grid(yVar ~ xVar, scales = "free") + 
    labs(x = "X Variable", y = "Y Variable")
}
pairs_plot(eda_set)
```

```{r}
pca_plot <- function(df, scale = F) {
  autoplot(prcomp(df, scale = scale), data = df, 
         loadings = TRUE, loadings.colour = 'blue',
         loadings.label = TRUE, loadings.label.size = 3)
}
pca_set <- eda_set %>% select(- SalePrice) %>% select_if(is.numeric) %>% drop_na
pca_plot(pca_set)
eda_pca <- prcomp(pca_set)
summary(eda_pca)
eda_pca %>% extract(c("sdev", "center")) %>% as.data.frame %>% mutate(Feature = rownames(.)) %>% arrange(desc(center)) %>% 
  slice(1:20) %>% 
  pull(Feature) -> important_features
print(important_features)
  
```



# Feature Selection

# A First Approach to Machine Learning: Linear Regression

```{r}
simple_model <- lm(data = training_set, formula = SalePrice ~ LotArea)
simple_model %>% augment %>% select(`SalePrice (Actual)` = SalePrice, `SalePrice (Predicted)` = .fitted, LotArea) %>%
  gather(Value, SalePrice, - LotArea) %>%
  ggplot(aes(LotArea, SalePrice, col = Value, linetype = Value)) + geom_point()

initial_model <- lm(data = training_set, formula = SalePrice ~ LotArea + GrLivArea)
initial_model %>% augment %>% select(`SalePrice (Actual)` = SalePrice, `SalePrice (Predicted)` = .fitted, LotArea, GrLivArea) %>%
  gather(Value, SalePrice, - c(LotArea, GrLivArea)) %>%
  ggplot(aes(LotArea, SalePrice, col = Value, linetype = Value)) + geom_point()
initial_model %>% augment %>% 
  ggplot(aes(LotArea, .resid)) + geom_point() + 
  geom_hline(yintercept = 0, col = "black", linetype = "dashed") +
  labs(x = "Lot Area (Square Ft.)", y = "Residual", title = "Residual Plot of Iniital Linear Model") 

model <- lm(data = training_set, 
            formula = as.formula(paste("SalePrice ~ ", paste(important_features %>% paste0("`", ., "`"), collapse= " + "))))
model %>% augment %>% 
  ggplot(aes(SalePrice, .resid)) + geom_point() + 
  geom_hline(yintercept = 0, col = "black", linetype = "dashed") +
  labs(x = "Sales Price", y = "Residual", title = "Residual Plot of 'Important Features' Linear Model") 
```




# Extensions to Linear Regression

# Conclusion

This ends our textbook-style primer into deep learning with Keras. While this was just an introduction to neural nets, we hope that you can now see some of the workflow patterns associated with machine learning. Feel free to play around with the code above to get a better feel for the hyperparameters of the neural net model. As always, please email [`contact@arun.run`](mailto:contact@arun.run) or [`prc@berkeley.edu`](mailto:prc@berkeley.edu) with any questions or concerns whatsoever. Happy machine learning!

## Sneakpeek at SUSA Kaggle Competition II

After Spring Break, we will be guiding you through a four-week collaborative Kaggle competition with your peers in Career Exploration! We want to give you the experience of working with real data, using real machine learning algorithms, in an educational setting. You will have to choose either Python or R, and dive into reading kernels on the Kaggle website, use visualization and feature engineering to improve your score, and maybe even pick up a few advanced deep learning models along the way. If this sounds a bit intimidating right now, do not fret! Your SUSA Mentors will be there to mentor you through the whole thing. So rest up during Spring Break, and come back ready to tackle your biggest data challenge yet!

# Additional Reading
* For more information on the Kaggle API, a command-line program used to download and manage Kaggle datasets, visit the [Kaggle API Github page](https://github.com/Kaggle/kaggle-api)  
* For an interactive guide to learning R and Python, visit [DataCamp](https://www.datacamp.com/) a paid tutorial website for learning data computing.
